### DISCUSSION TITLE
Extracting path and inserting cells along path points

### DISCUSSION THREAD

**Discussion by Qvadi** [iqm-finland/KQCircuits#72] [Q&A]
Hi KQC,
I was wondering if its possible to extract a path for a meandering waveguide_composite element and convert that path to a set of points which I can use to insert elements (for eg airbridges)? I see that this has been done for the spiral resonator element using a similar logic using get_angle and other functions. 

I had something simpler (straight waveguide) in mind to play around:

```
#Two points to define the waveguide
in_p1 = (0, 0)
out_p2 = (1000, 0)

#Define node list
nodes_probe_path = [Node(in_p1), Node(out_p2)]

#Define path using Param
probe_path = Param(pdt.TypeShape, "Probe path", pya.DPath([pya.DPoint(in_p1[0], in_p1[1]), pya.DPoint(out_p2[0], out_p2[1])],10))

#Insert the waveguide using waveguide_composite element        
probeline = self.insert_cell(WaveguideComposite, nodes = nodes_probe_path, a=self.a, b=self.b)

#Extract the points in the path 
probepoints = list(probe_path.each_point()) #------ Here, I get an error with each_point() saying probeline has no attribute each_point

#If I am able to extract the points as a list, then a simple loop to add bridges along the path
for i in len(probepoints):
      self.insert_cell(AirbridgeMultiFace, probepoints(i))

```
For a meander, it would follow the same logic but with a bit more code on calculating angles and so on....

**Comment by qpavsmi** [comment]
Hello,

To use `Param` objects purposefully, you need to define those as member variables of a class. So for your custom element you need to wrap it as a class that derives `Element` (e.g. `MyElement(Element)`) and it needs to have a `build()` function. Then when you list `probe_path` as `Param`, you can use `probe_path` in the `build` function as a `pya.DPath` object.

If you want to see you code snippet work quickly, simply replace the `probe_path` definition with `probe_path = pya.DPath([pya.DPoint(in_p1[0], in_p1[1]), pya.DPoint(out_p2[0], out_p2[1])],10)`

You should also consider using the features of `WaveguideComposite` in order to achieve your goal to insert airbridges, if you're not planning using this element as a child of another `WaveguideComposite` element. Would something like this be sufficient for you?

![image](https://github.com/iqm-finland/KQCircuits/assets/109152229/5fa97511-1a92-4fa2-8df7-281df827e14a)

This is achieved by running this piece of code as macro:

```
from kqcircuits.pya_resolver import pya
from kqcircuits.klayout_view import KLayoutView

from kqcircuits.elements.waveguide_composite import WaveguideComposite
from kqcircuits.elements.waveguide_composite import Node

view = KLayoutView()
layout = view.layout

# WaveguideComposite with airbridges
view.insert_cell(WaveguideComposite, nodes=[Node((0.0, 0.0)), Node((1500.0, 0.0), length_before=3000.0, n_bridges=12)])
```
